src/api/app.py

```diff
@@ -140,49 +153,102 @@ def session_id(api_key: str) -> dict[str, int]:
 
     Returns:
         {"closed_successfully": True or False}
     """
-    try:
-        demo_manager.close(session_id)
-    except Exception:
-        ...
-
-    return {"closed_successfully": ...}
-
-
-@websocket_listener("/demos")
-async def demo_session(data: dict[str, str | bytes | int]) -> dict[str, str]:
-    """Handle incoming data from a client upload.
-
-    Smart enough to know where to write data to based on the session ID
-    as to handle reconnecting/duplicates.
-
-    Might want to implement something like
-    https://docs.litestar.dev/2/usage/websockets.html#class-based-websocket-handling
-    because it looks cleaner and likely can handle auth/accepting/valid session id better
-    """
-    demo_manager.handle_demo_data(data)
+    engine = request.app.state.engine
+    current_time = datetime.now().astimezone(timezone.utc).isoformat()
+    with engine.connect() as conn:
+        conn.execute(
+            sa.text(
+                "UPDATE demo_sessions SET active = False, end_time = :end_time, updated_at = :updated_at WHERE session_id = :session_id AND api_key = :api_key;"  # noqa
+            ),
+            {
+                "api_key": api_key,
+                "session_id": session_id,
+                "end_time": current_time,
+                "updated_at": current_time,
+            },
+        )
+        conn.commit()
+
+    return {"closed_successfully": True}
+
+
+class DemoHandler(WebsocketListener):
+    path = "/demos"
+    receive_mode = "binary"
+
+    async def on_accept(self, socket: WebSocket, api_key: str, session_id: str) -> None:
+        engine = socket.app.state.async_engine
+        exists = await _check_key_exists(engine, api_key)
+        if not exists:
+            await socket.close()
+
+        active = await _check_is_active(engine, api_key, session_id)
+        if not active:
+            await socket.close()
+
+        self.api_key = api_key
+        self.session_id = session_id
+        self.path = os.path.join(DEMOS_PATH, f"{session_id}.dem")
+        self.handle = open(os.path.join(DEMOS_PATH, f"{session_id}.dem"), "wb")
+
+    def on_disconnect(self, socket: WebSocket) -> None:
+        self.handle.close()
+
+        demo = open(self.path, "rb").read()
+
+        current_time = datetime.now().astimezone(timezone.utc).isoformat()
+        engine = socket.app.state.engine
+
+        with engine.connect() as conn:
+            conn.execute(
+                sa.text(
+                    "UPDATE demo_sessions SET active = False, end_time = :end_time, demo = :demo, updated_at = :updated_at WHERE session_id = :session_id AND api_key = :api_key;"  # noqa
+                ),
+                {
+                    "api_key": self.api_key,
+                    "session_id": self.session_id,
+                    "end_time": current_time,
+                    "updated_at": current_time,
+                    "demo": demo,
+                },
+            )
+            conn.commit()
```

Comment:
Does this mean you will continue to store all the demo sessions that happen? Or will they be automatically deleted once the app is closed? If they persist, it might be good to change this such that the record is simply deleted rather then active set to false. Is it possible a demo session would get re-used? I was assuming not since each session has a unique ID.

Reply:
The workflow is that a user could have many demo sessions, but only one active one at a time. I think planning additional workflow that will "retire" non-active demo sessions on a schedule to a more warehouse-style solution